homeassistant/components/thethingsnetwork/config_flow.py

```diff
@@ -0,0 +1,118 @@
+"""The Things Network's integration config flow."""
+
+from collections.abc import Mapping
+import logging
+import traceback
+from typing import Any
+
+from ttn_client import TTNAuthError, TTNClient
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+
+from .const import CONF_API_KEY, CONF_APP_ID, CONF_HOSTNAME, DOMAIN, TTN_API_HOSTNAME
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TTNFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize flow."""
+        self._user_input: Mapping[str, Any] = {}
+        self._user_input[CONF_HOSTNAME] = TTN_API_HOSTNAME
+        self._reauth_entry: ConfigEntry | None = None
+
+    @property
+    def schema(self) -> vol.Schema:
+        """Return current schema."""
+
+        return vol.Schema(
+            {
+                vol.Required(
+                    CONF_HOSTNAME, default=self._user_input[CONF_HOSTNAME]
+                ): str,
+                vol.Required(
+                    CONF_APP_ID, default=self._user_input.get(CONF_APP_ID)
+                ): str,
+                vol.Required(
+                    CONF_API_KEY, default=self._user_input.get(CONF_API_KEY)
+                ): str,
+            }
+        )
```

Comment:
The only step that actually gets the user_input is the async_step_user, so I don't really see a point in storing the user_input in a config flow variable

Reply:
But that can just happen in the `async_setup_user`, since in an invalid case, you already have everything stored in `user_input`, we even have a nice helper for that

Reply:
removed self._user_input - and added the password selector.